File: src/main/java/com/meetify/server/configs/DatabaseConfig.java
package com.meetify.server.configs;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.env.Environment;
import org.springframework.dao.annotation.PersistenceExceptionTranslationPostProcessor;
import org.springframework.jdbc.datasource.DriverManagerDataSource;
import org.springframework.orm.jpa.JpaTransactionManager;
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import javax.sql.DataSource;
import java.util.Properties;

/**
 * com.meetify.server.configs
 * Created by kr3v on 04.10.2016.
 */

@SuppressWarnings("SpringAutowiredFieldsWarningInspection")
@Configuration
@EnableTransactionManagement
public class DatabaseConfig {
    
    
    @Autowired
    private Environment env;
    @Autowired
    private DataSource dataSource;
    
    @SuppressWarnings("SpringJavaAutowiringInspection")
    @Autowired
    private LocalContainerEntityManagerFactoryBean entityManagerFactory;
    
    @Bean
    public DataSource dataSource() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource();
        dataSource.setDriverClassName(env.getProperty("db.driver"));
        dataSource.setUrl(env.getProperty("db.url"));
        dataSource.setUsername(env.getProperty("db.username"));
        dataSource.setPassword(env.getProperty("db.password"));
        return dataSource;
    }
    
    /**
     * Declare the JPA entity manager factory.
     *
     * @return LocalContainerEntityManagerFactoryBean
     */
    @Bean
    public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
        LocalContainerEntityManagerFactoryBean entityManagerFactory =
                new LocalContainerEntityManagerFactoryBean();
    
        entityManagerFactory.setDataSource(dataSource);
    
        // Classpath scanning of @Component, @Service, etc annotated class
        entityManagerFactory.setPackagesToScan(
                env.getProperty("entitymanager.packagesToScan"));
    
        // Vendor adapter
        HibernateJpaVendorAdapter vendorAdapter = new HibernateJpaVendorAdapter();
        entityManagerFactory.setJpaVendorAdapter(vendorAdapter);
    
        // Hibernate properties
        Properties additionalProperties = new Properties();
        additionalProperties.put(
                "hibernate.dialect",
                env.getProperty("hibernate.dialect"));
        additionalProperties.put(
                "hibernate.show_sql",
                env.getProperty("hibernate.show_sql"));
        additionalProperties.put(
                "hibernate.hbm2ddl.auto",
                env.getProperty("hibernate.hbm2ddl.auto"));
        entityManagerFactory.setJpaProperties(additionalProperties);
    
        return entityManagerFactory;
    }
    
    /**
     * Declare the transaction manager.
     *
     * @return JpaTransactionManager
     */
    @Bean
    public JpaTransactionManager transactionManager() {
        JpaTransactionManager transactionManager =
                new JpaTransactionManager();
        transactionManager.setEntityManagerFactory(
                entityManagerFactory.getObject());
        return transactionManager;
    }
    
    /**
     * PersistenceExceptionTranslationPostProcessor is a bean post processor
     * which adds an advisor to any bean annotated with Repository so that any
     * platform-specific exceptions are caught and then rethrown as one
     * Spring's unchecked data access exceptions (i.e. a subclass of
     * DataAccessException).
     *
     * @return PersistenceExceptionTranslationPostProcessor
     */
    @Bean
    public PersistenceExceptionTranslationPostProcessor exceptionTranslation() {
        return new PersistenceExceptionTranslationPostProcessor();
    }
}

File: src/main/java/com/meetify/server/configs/DatabaseConfig.kt
package com.meetify.server.configs

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.Environment
import org.springframework.dao.annotation.PersistenceExceptionTranslationPostProcessor
import org.springframework.jdbc.datasource.DriverManagerDataSource
import org.springframework.orm.jpa.JpaTransactionManager
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter
import org.springframework.transaction.annotation.EnableTransactionManagement
import java.util.*
import javax.sql.DataSource

/**
 * Клас, який використовується самим фреймворком Spring для конфігурації Hibernate.
 * @author  Дмитро Байнак
 * @version 0.0.1
 * @since   0.0.1
 * @property    env                     поле, що дозволяє завантажувати конфігурацію, задану в application.properties.
 * @property    dataSource              поле, що надає доступу до бази даних.
 * @property    entityManagerFactory    фабрика, що дозволяє створити екземляри entityManager.
 */
@Configuration @EnableTransactionManagement @SuppressWarnings("SpringAutowiredFieldsWarningInspection")
open class DatabaseConfig {

    @Autowired
    private val env: Environment? = null

    @Autowired
    private val dataSource: DataSource? = null

    @Suppress("SpringKotlinAutowiring") @Autowired
    private val entityManagerFactory: LocalContainerEntityManagerFactoryBean? = null

    /**
     * Метод, що дозволяє створити DataSource з параметрами, заданими в application.properties
     * @return dataSource
     */
    @Bean
    open fun dataSource(): DataSource = DriverManagerDataSource().apply {
        setDriverClassName(env!!.getProperty("db.driver"))
        url = env.getProperty("db.url")
        username = env.getProperty("db.username")
        password = env.getProperty("db.password")
    }

    /**
     * Метод, що дозволяє створити фабрику диспетчерів сутностей.
     * @return LocalContainerEntityManagerFactoryBean
     */
    @Bean
    open fun entityManagerFactory(): LocalContainerEntityManagerFactoryBean {
        return LocalContainerEntityManagerFactoryBean().apply {
            this.dataSource = this@DatabaseConfig.dataSource!!
            setPackagesToScan(env!!.getProperty("entitymanager.packagesToScan"))
            jpaVendorAdapter = HibernateJpaVendorAdapter()
            setJpaProperties(Properties().apply {
                put("hibernate.dialect", env.getProperty("hibernate.dialect"))
                put("hibernate.show_sql", env.getProperty("hibernate.show_sql"))
                put("hibernate.hbm2ddl.auto", env.getProperty("hibernate.hbm2ddl.auto"))
            })
        }
    }

    /**
     * Метод, що дозволяє створити менеджер транзакцій.
     * @return JpaTransactionManager
     */
    @Bean
    open fun transactionManager(): JpaTransactionManager {
        return JpaTransactionManager().apply {
            entityManagerFactory = this@DatabaseConfig.entityManagerFactory!!.`object`
        }
    }

    /**
     * PersistenceExceptionTranslationPostProcessor це компонентний пост-процесор,
     * який дає змоги будь-якій компоненті із аннотацією Repository
     * "кидати" специфічні відловлювані винятки і передавати їх фреймворкові.
     * @return PersistenceExceptionTranslationPostProcessor
     */
    @Bean
    open fun exceptionTranslation(): PersistenceExceptionTranslationPostProcessor {
        return PersistenceExceptionTranslationPostProcessor()
    }
}
